py/server/deephaven/table_listener.py

```diff
@@ -393,3 +391,199 @@ def listen(t: Table, listener: Union[Callable, TableListener], description: str
+                the listener is safe, it is not recommended because reading or operating on the result tables of those
+                operations may not be safe. It is best to perform the operations on the dependent tables beforehand,
+                and then add the result tables as dependencies to the listener so that they can be safely read in it.
+
+        Raises:
+            DHError
+        """
+        if len(tables) < 2:
+            raise DHError(message="A merged listener must listen to at least 2 refreshing tables.")
+
+        self.tables = tables
+        self.listener_recorders = [_ListenerRecorder(t) for t in tables]
+
+        self.dependencies = dependencies
+
+        if isinstance(listener, MergedListener):
+            self.listener = listener.on_update
+        else:
+            self.listener = listener
+        n_params = len(signature(self.listener).parameters)
+        if n_params != 2:
+            raise ValueError("merged listener function must have 2 parameters (updates, is_replay).")
+
+
+        try:
+            self.merged_listener_adapter = _JPythonMergedListenerAdapter.create(
+                        to_sequence(self.listener_recorders),
+                        to_sequence(self.dependencies),
+                        description,
+                        self)
+            self.started = False
+        except Exception as e:
+            raise DHError(e, "failed to create a merged listener adapter.") from e
+
+    def _process(self, replay_updates: Optional[jpy.JType] = None) -> None:
+        """Process the table updates from the listener recorders.
+
+        Args:
+            replay_updates (Optional[jpy.JType]): the replay updates, only provided during replay.
+        """
+        if replay_updates:
+            self.listener({t: TableUpdate(t, tu) for t, tu in zip(self.tables, j_list_to_list(replay_updates))}, True)
+        else:
+            self.listener({lr.table: lr.table_update() for lr in self.listener_recorders}, False)
+
+    def start(self, do_replay: bool = False, replay_lock: Literal["shared", "exclusive"] = "shared") -> None:
+        """Start the listener by registering it with the tables and listening for updates.
+
+        Args:
+            do_replay (bool): whether to replay the initial snapshots of the tables, default is False
+            replay_lock (str): the lock type used during replay, default is 'shared', can also be 'exclusive'.
+
+        Raises:
+            DHError
+        """
+        if self.started:
+            raise RuntimeError("Attempting to start an already started merged listener..")
+
+        try:
+            if do_replay:
+                _do_locked(self.tables[0].update_graph, self.merged_listener_adapter.replay, lock_type=replay_lock)
+
+            with update_graph.auto_locking_ctx(self.tables[0].update_graph):
+                for lr in self.listener_recorders:
+                    lr.table.j_table.addUpdateListener(lr.j_listener_recorder)
```

Comment:
The change you've made is incorrect; the lock must be held for the duration of the replay *and* the subscription. Here's my intended simplification.

Note that there's no `replay_lock` in use: we should remove that argument, it's an unnecessary complication. The user can simply acquire the exclusive lock themselves if they need that kind of control, which they almost certainly will _not_ (because they _cannot_ be doing anything in their listener that actually requires holding the exclusive lock, i.e. using a condition variable). See `io.deephaven.engine.updategraph.UpdateGraphLock#checkForIllegalLockFromRefreshThread`: you can't get a lock from a refresh thread, which means you will never have the exclusive lock in your listener.
```suggestion
            with update_graph.auto_locking_ctx(self.tables[0].update_graph):
                if do_replay:
                    self.merged_listener_adapter.replay()
                for lr in self.listener_recorders:
                    lr.table.j_table.addUpdateListener(lr.j_listener_recorder)
```